### PR TITLE
Compute index in merkle tree

### LINKED ISSUE #8844: Compute index/position in Merkle tree correctly
### Expected behavior
Inside message recovery modules, indexes of Proof should remove the preceding `1` before comparing.

### Actual behavior
preceding `1` not removed

Additional Read: https://github.com/LiskHQ/lips/pull/445

### PR DESCRIPTION
**Pull request by Phanco**
### What was the problem?

This PR resolves #8844

### How was it solved?
Additional operation when comparing indices

### How was it tested?
Updated unit tests

### CODE REVIEW

File: framework/src/modules/interoperability/mainchain/commands/recover_message.ts [hunk 1]
```diff
@@ -91,9 +91,11 @@ export class RecoverMessageCommand extends BaseInteroperabilityCommand<Mainchain
...
-		// as the idxs are sorted in ascending order.
-		if (idxs[0] < terminatedOutboxAccount.partnerChainInboxSize) {
+		// as the idxs are sorted in ascending order. Note that one must unset the most significant
+		// bit of an encoded index in idxs in order to get the position in the tree.
+		// See https://github.com/LiskHQ/lips/blob/main/proposals/lip-0031.md#proof-serialization.
+		const firstPosition = parseInt(idxs[0].toString(2).slice(1), 2);
+		if (firstPosition < terminatedOutboxAccount.partnerChainInboxSize) {
 			return {
 				status: VerifyStatus.FAIL,
 				error: new Error('Cross-chain message is not pending.'),
```
Comment (sergeyshemyakov) [inline framework/src/modules/interoperability/mainchain/commands/recover_message.ts:101]: I think you miss the second check from the LIP: https://github.com/LiskHQ/lips/blob/c24059afc77eebf8288faf49438618c75ea82da6/proposals/lip-0054.md?plain=1#L400

**Review by sergeyshemyakov — CHANGES_REQUESTED** [review]
Just one comment from my side, otherwise looks good

File: framework/src/modules/interoperability/mainchain/commands/recover_message.ts (hunk 1 above)
Comment (Phanco) [reply framework/src/modules/interoperability/mainchain/commands/recover_message.ts:101]: The second check will be handled by this PR 👍 
https://github.com/LiskHQ/lisk-sdk/pull/8837

File: framework/src/modules/interoperability/mainchain/commands/recover_message.ts [hunk 2]
```diff
@@ -91,9 +91,11 @@ export class RecoverMessageCommand extends BaseInteroperabilityCommand<Mainchain
 			}
 		}
 
-		// Check that the CCMs are still pending. We can check only the first one,
-		// as the idxs are sorted in ascending order.
-		if (idxs[0] < terminatedOutboxAccount.partnerChainInboxSize) {
+		// as the idxs are sorted in ascending order. Note that one must unset the most significant
+		// bit of an encoded index in idxs in order to get the position in the tree.
+		// See https://github.com/LiskHQ/lips/blob/main/proposals/lip-0031.md#proof-serialization.
+		const firstPosition = parseInt(idxs[0].toString(2).slice(1), 2);
```
Comment (AndreasKendziorra) [inline framework/src/modules/interoperability/mainchain/commands/recover_message.ts:97]: sorry to jump in here, but it's related to the other PR I'm currently reviewing. We should have a unit test for the edge case `idxs = [0]`, i.e. `idxs[0]`does not follow the format of having a prepended `1`. The verify function should fail in this case.

**Review by AndreasKendziorra — CHANGES_REQUESTED** [review]

File: framework/src/modules/interoperability/mainchain/commands/recover_message.ts (hunk 2 above)
Comment (Phanco) [reply framework/src/modules/interoperability/mainchain/commands/recover_message.ts:104]: Updated

File: framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts [hunk 3]
```diff
@@ -59,7 +59,13 @@ import {
 import { CcmSendSuccessEvent } from '../../../../../../src/modules/interoperability/events/ccm_send_success';
 import { InvalidRMTVerification } from '../../../../../../src/modules/interoperability/events/invalid_rmt_verification';
 
-const getHeight = (size: number) => Math.ceil(Math.log2(size)) + 1;
+// Choose the index such that the position of the ccm in the outbox tree is at terminatedChainOutboxSize,
+// i.e. just outside of the tree. Note that a 1 has to be prepended to the binary representation of the index, which is done by adding 2**indexLength.
+// See https://github.com/LiskHQ/lips/blob/main/proposals/lip-0031.md#proof-serialization.
+const appendPrecedingToIndices = (indices: number[], terminatedChainOutboxSize: number) => {
```
Comment (mitsuaki-u) [inline framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts:65]: We can either move this inside the first level describe, or move the two createContext functions where this is, to keep it consistent

**Review by mitsuaki-u — APPROVED** [review]

File: framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts (hunk 3 above)
Comment (Phanco) [reply framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts]: Updated

File: framework/src/modules/interoperability/mainchain/commands/recover_message.ts [hunk 4]
```diff
@@ -92,9 +92,18 @@ export class RecoverMessageCommand extends BaseInteroperabilityCommand<Mainchain
 			}
 		}
 
-		// Check that the CCMs are still pending. We can check only the first one,
```
Comment (AndreasKendziorra) [inline framework/src/modules/interoperability/mainchain/commands/recover_message.ts:95]: It seems that this line got accidentally lost?

File: framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts [hunk 5]
```diff
@@ -59,16 +59,18 @@ import {
...
-
 describe('MessageRecoveryCommand', () => {
 	const interopModule = new MainchainInteroperabilityModule();
 	const leafPrefix = Buffer.from([0]);
+
+	// Choose the index such that the position of the ccm in the outbox tree is at terminatedChainOutboxSize,
+	// i.e. just outside of the tree. Note that a 1 has to be prepended to the binary representation of the index, which is done by adding 2**indexLength.
+	// See https://github.com/LiskHQ/lips/blob/main/proposals/lip-0031.md#proof-serialization.
+	const appendPrecedingToIndices = (indices: number[], terminatedChainOutboxSize: number) => {
+		const mostSignificantBit = 2 ** (Math.ceil(Math.log2(terminatedChainOutboxSize)) + 1);
```
Comment (AndreasKendziorra) [inline framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts:70]: I find the name `mostSignificantBit` a bit misleading. Instead, `(Math.ceil(Math.log2(terminatedChainOutboxSize)) + 1)` describes the most significant bit. However, this is the most significant bit of the result and not of the input (`terminatedChainOutboxSize`). So, It would maybe still be confusing to call this value the "most significant bit".

File: framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts [hunk 6]
```diff
@@ -59,16 +59,18 @@ import {
...
-const appendPrecedingToIndices = (indices: number[], terminatedChainOutboxSize: number) => {
-	const mostSignificantBit = 2 ** terminatedChainOutboxSize;
-	return indices.map(index => index + mostSignificantBit);
-};
-
 describe('MessageRecoveryCommand', () => {
 	const interopModule = new MainchainInteroperabilityModule();
 	const leafPrefix = Buffer.from([0]);
+
+	// Choose the index such that the position of the ccm in the outbox tree is at terminatedChainOutboxSize,
```
Comment (AndreasKendziorra) [inline framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts:66]: > Choose the index such that the position of the ccm in the outbox tree is at terminatedChainOutboxSize, i.e. just outside of the tree.

This sentence does not fit in general to this function. It only makes sense for it's application in line 329.

File: framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts [hunk 7]
```diff
@@ -59,16 +59,18 @@ import {
...
+	// i.e. just outside of the tree. Note that a 1 has to be prepended to the binary representation of the index, which is done by adding 2**indexLength.
```
Comment (AndreasKendziorra) [inline framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts:67]: > which is done by adding 2**indexLength.

the current code does not use the name `indexLength`.

**Review by AndreasKendziorra — CHANGES_REQUESTED** [review]

**Review by sergeyshemyakov — APPROVED** [review]

File: framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts (hunk 5 above)
Comment (sergeyshemyakov) [reply framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts:70]: I think the naming is fine, especially since it is just in tests.

File: framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts (hunk 7 above)
Comment (ishantiw) [reply framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts]: Comments moved to L324 and it has been renamed to `index.length`

File: framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts (hunk 6 above)
Comment (ishantiw) [reply framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts]: Moved it to L324

File: framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts (hunk 5 above)
Comment (ishantiw) [reply framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts]: used shorthand for the function so don't have any var name anymore

File: framework/src/modules/interoperability/mainchain/commands/recover_message.ts (hunk 4 above)
Comment (ishantiw) [reply framework/src/modules/interoperability/mainchain/commands/recover_message.ts:95]: added it back

File: framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts [hunk 8]
```diff
@@ -326,6 +321,9 @@ describe('MessageRecoveryCommand', () => {
 			];
 			ccmsEncoded = ccms.map(ccm => codec.encode(ccmSchema, ccm));
 			transactionParams.crossChainMessages = [...ccmsEncoded];
+			// Choose the index such that the position of the ccm in the outbox tree is at terminatedChainOutboxSize,
+			// i.e. just outside of the tree. Note that a 1 has to be prepended to the binary representation of the index, which is done by adding 2**(index.length).
+			// See https://github.com/LiskHQ/lips/blob/main/proposals/lip-0031.md#proof-serialization.
```
Comment (AndreasKendziorra) [inline framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts:326]: > Note that a 1 has to be prepended to the binary representation of the index, which is done by adding 2**(index.length). See https://github.com/LiskHQ/lips/blob/main/proposals/lip-0031.md#proof-serialization.

I would maybe remove this part of the comment as this is not specific to this test, or move it back to line 65. If you prefer to keep it here, I would at least remove "which is done by adding 2**(index.length)" because it's probably not clear to the reader what `index.length` means.

File: framework/src/modules/interoperability/mainchain/commands/recover_message.ts [hunk 9]
```diff
@@ -92,7 +92,7 @@ export class RecoverMessageCommand extends BaseInteroperabilityCommand<Mainchain
 			}
 		}
 
-		// as the idxs are sorted in ascending order. Note that one must unset the most significant
+		// As the idxs are sorted in ascending order. Note that one must unset the most significant
 		// bit of an encoded index in idxs in order to get the position in the tree.
 		// See https://github.com/LiskHQ/lips/blob/main/proposals/lip-0031.md#proof-serialization.
```
Comment (AndreasKendziorra) [inline framework/src/modules/interoperability/mainchain/commands/recover_message.ts:97]:
```suggestion
		// Encoded indices equal to zero are also not valid here as they represent non-existing
		// entries. It's sufficient to check only the first one due the ascending order.
		// See https://github.com/LiskHQ/lips/blob/main/proposals/lip-0031.md#proof-serialization.
```

File: framework/src/modules/interoperability/mainchain/commands/recover_message.ts [hunk 10]
```diff
@@ -102,6 +102,8 @@ export class RecoverMessageCommand extends BaseInteroperabilityCommand<Mainchain
 			};
 		}
 
+		// Check that the CCMs are still pending. We can check only the first one,
+		// as the idxs are sorted in ascending order.
```
Comment (AndreasKendziorra) [inline framework/src/modules/interoperability/mainchain/commands/recover_message.ts:106]:
```suggestion
		// Check that the CCMs are still pending. We can check only the first one,
		// as the idxs are sorted in ascending order. Note that one must unset the most significant
    	        // bit a of an encoded index in idxs in order to get the position in the tree.
		// See https://github.com/LiskHQ/lips/blob/main/proposals/lip-0031.md#proof-serialization.
```

**Review by AndreasKendziorra — APPROVED** [review]
Some comments are still a bit messed up and could be improved as suggested. But otherwise looks good.

File: framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts (hunk 8 above)
Comment (ishantiw) [reply framework/test/unit/modules/interoperability/mainchain/commands/recover_message.spec.ts:326]: yeah doesn't suit here, I will remove it as its not necessary to put comments in the test and also the comment doesn't really help much here. I have removed this part of the comment to avoid any confusion
